File: api/server/src/main/java/su/plo/voice/api/server/event/command/CommandsRegisterEvent.java
package su.plo.voice.api.server.event.command;

import lombok.Getter;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import su.plo.lib.api.server.command.MinecraftCommand;
import su.plo.lib.api.server.command.MinecraftCommandManager;
import su.plo.voice.api.addon.AddonManager;
import su.plo.voice.api.event.Event;
import su.plo.voice.api.server.PlasmoVoiceServer;

/**
 * Called when the server is ready to register commands.
 * <br/>
 * Will not be called if addon was registered by {@link AddonManager#load(Object)}
 */
@RequiredArgsConstructor
public final class CommandsRegisterEvent implements Event {

    @Getter
    private final @NonNull PlasmoVoiceServer voiceServer;
    @Getter
    private final @NonNull MinecraftCommandManager<MinecraftCommand> commandManager;
}

File: api/server-common/src/main/kotlin/su/plo/lib/api/server/event/command/CommandsRegisterEvent.kt
package su.plo.lib.api.server.event.command

import su.plo.lib.api.event.MinecraftGlobalEvent
import su.plo.lib.api.server.MinecraftCommonServerLib
import su.plo.lib.api.server.command.MinecraftCommand
import su.plo.lib.api.server.command.MinecraftCommandManager
import su.plo.lib.api.server.event.command.CommandsRegisterEvent.Callback

abstract class CommandsRegisterEvent<C : MinecraftCommand, S : MinecraftCommonServerLib>
    : MinecraftGlobalEvent<Callback<C, S>>(
    { callbacks ->
        Callback { commandManager, minecraftServer ->
            callbacks.forEach { callback -> callback.onCommandsRegister(commandManager, minecraftServer) }
        }
    }
) {
    fun interface Callback<C : MinecraftCommand, S : MinecraftCommonServerLib> {

        fun onCommandsRegister(commandManager: MinecraftCommandManager<C>, minecraftServer: S)
    }
}
